kayenta-wavefront/src/main/java/com/netflix/kayenta/wavefront/metrics/WavefrontMetricsService.java

```diff
@@ -0,0 +1,141 @@
+import java.io.IOException;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+@Builder
+@Slf4j
+public class WavefrontMetricsService implements MetricsService {
+
+    @NotNull
+    @Singular
+    @Getter
+    private List<String> accountNames;
+
+    @Autowired
+    private final AccountCredentialsRepository accountCredentialsRepository;
+
+    @Autowired
+    private final Registry registry;
+
+    @Override
+    public String getType() {
+        return "wavefront";
+    }
+
+    @Override
+    public boolean servicesAccount(String accountName) {
+        return accountNames.contains(accountName);
+    }
+
+    @Override
+    public String buildQuery(String metricsAccountName,
+                             CanaryConfig canaryConfig,
+                             CanaryMetricConfig canaryMetricConfig,
+                             CanaryScope canaryScope) {
+        WavefrontCanaryMetricSetQueryConfig queryConfig = (WavefrontCanaryMetricSetQueryConfig)canaryMetricConfig.getQuery();
+
+        String query = queryConfig.getMetricName();
+        if (canaryScope.getScope() != null && !canaryScope.getScope().equals("")) {
+            query = query + ", " + canaryScope.getScope();
+        }
+
+        query = "ts(" + query + ")";
+        if (queryConfig.getAggregate() != null && !queryConfig.getAggregate().equals("")) {
+            query = queryConfig.getAggregate() + "(" + query + ")";
+        }
+
+        return query;
+    }
+
+    @Override
+    public List<MetricSet> queryMetrics(String accountName,
+                                        CanaryConfig canaryConfig,
+                                        CanaryMetricConfig canaryMetricConfig,
+                                        CanaryScope canaryScope) throws IOException {
+        WavefrontCanaryScope wavefrontCanaryScope = (WavefrontCanaryScope) canaryScope;
+        WavefrontNamedAccountCredentials accountCredentials = (WavefrontNamedAccountCredentials) accountCredentialsRepository
+                .getOne(accountName)
+                .orElseThrow(() -> new IllegalArgumentException("Unable to resolve account " + accountName + "."));
+        WavefrontCredentials credentials = accountCredentials.getCredentials();
+        WavefrontRemoteService wavefrontRemoteService = accountCredentials.getWavefrontRemoteService();
+
+        String query = buildQuery(accountName,
+                canaryConfig,
+                canaryMetricConfig,
+                canaryScope);
+
+        WavefrontTimeSeries timeSeries = wavefrontRemoteService.fetch(
+                credentials.getApiToken(),
+                "Kayenta-Query",
+                query,
+                wavefrontCanaryScope.getStart().toEpochMilli(),
+                wavefrontCanaryScope.getEnd().toEpochMilli(),
+                wavefrontCanaryScope.getGranularity(),
+                "MEAN",
```

Comment:
Does Wavefront support other values for the summarization method?

Reply:
@fieldju Yes it supports Mean, Median, Min, Max, Sum, Count, First, and Last.  If there are multiple values within the granularity for a metric (i.e. a metric having a data points every second but the query requests a minute granularity). I just pushed a change to allow users to define which summarization they would prefer.